dashboard.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+import collections
+import glob
+import importlib
+import inspect
+import os
+import pathlib
+import sys
+
+import altair
+import pandas
+import panel
+import sqlalchemy
+from bokeh.resources import INLINE
+
+panel.extension("vega")
+
+
+def get_test_source():
+    """
+    Crawl the tests directory and try to grab code for each test on a best-effort
+    basis. This relies on the tests being importable from this script, so the
+    environment should be similar enough that that is possible.
+    """
+    source: dict[str, str] = {}
+    files = glob.glob("tests/**/test_*.py", recursive=True)
+    for f in files:
+        try:
+            # Fragile!
+            mod = importlib.import_module(f.replace("/", ".")[: -len(".py")])
+            tests = [a for a in dir(mod) if a.startswith("test_")]
+            for test in tests:
+                if fn := getattr(mod, test, None):
+                    if not callable(fn):
+                        continue
+                    source[f[len("tests/") :] + "::" + test] = inspect.getsource(fn)
+        except Exception:
+            pass
+    return source
+
+
+source = get_test_source()
+
+
+def make_timeseries(originalname, df, spec) -> altair.Chart | None:
+    """
+    Make a single timeseries altair chart for a given test.
+
+    originalname: str
+        The name of the test without any fixture or other modifications.
+
+    df: pandas.DataFrame
+        A dataframe with the test data in it.
+
+    spec: ChartSpec
+        Data for how to render the timeseries
+    """
+    df = df[~df[spec.field].isna() & ~df.start.isna()]
```

Comment:
what's the purpose of this line?

Reply:
It's just dropping nulls for the relevant columns we are plotting. I should probably use `dropna` here.